modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/internal/AbstractArtifactClassLoaderBuilder.java

```diff
@@ -144,6 +164,35 @@ public ArtifactClassLoader build() throws IOException {
     return artifactClassLoader;
   }
 
+  /**
+   * Goes over the elements in the {@code pluginDescriptors} collection looking if it hasn't been resolved yet. If it hasn't
+   * then it looks it up through the {@link DependenciesProvider} and then creates an {@link ArtifactPluginDescriptor}.
+   *
+   * @param pluginDescriptors plugins to validate.
+   * @param visited plugins that are already resolved (by either the container or application initially, or by the resolver).
+   * @return the plugins that were obtained initially plus all the ones that were found.
+   * @throws DeploymentException if any dependency wasn't found properly
+   */
+  private List<ArtifactPluginDescriptor> getArtifactsWithDependencies(List<ArtifactPluginDescriptor> pluginDescriptors,
```

Comment:
Makes more sense to me to have this logic in the PluginDependenciesResolver implementation